user/tests/wiring/ble/heartrate_c/application.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+ ******************************************************************************
+  Copyright (c) 2015 Particle Industries, Inc.  All rights reserved.
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation, either
+  version 3 of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this program; if not, see <http://www.gnu.org/licenses/>.
+  ******************************************************************************
+ */
+
+/* Includes ------------------------------------------------------------------*/
+#include "application.h"
+
+BLEScanResult results[10];
+
+BLEAttribute heartrate;
+BLEDevice peer;
+
+void heartrateUpdated(uint8_t* data, uint16_t len) {
+
+}
+
+void setup() {
+}
+
+void loop() {
+    if (BLE.connected()) {
+        uint8_t newHr[4];
+        uint16_t len = 4;
+        heartrate->getValue(newHr, &len);

```

Comment:
I don't think we should have a separate C API. If there's some reason to keep the C API that I don't currently understand, then this second argument should not be a reference. Just make it a `size_t` passed by value.

Reply:
I have defined three kinds of `getValue()`:
```
    int getValue(uint8_t* buf, uint16_t* len) const;
    int getValue(String& str) const;

    template<typename T> int getValue(T* val) const {
        uint16_t len = sizeof(T);
        return getValue(static_cast<uint8_t*>(val), &len);
    }
```